### ISSUE TITLE
Breaking BC changes in the product list block

### ISSUE THREAD

**Issue by tzyganu** [magento/magento2#32268]
### Summary (*)
The `\Magento\Catalog\Block\Product\ListProduct::getLoadedProductCollection` is breaking BC (feel free to challenge it) in the 2.4 version.
In 2.4 the method looks like this:

```
    public function getLoadedProductCollection()
    {
        $collection = $this->_getProductCollection();

        $categoryId = $this->getLayer()->getCurrentCategory()->getId();
        foreach ($collection as $product) {
            $product->setData('category_id', $categoryId);
        }

        return $collection;
    }
```

In 2.3 and before it looked like this:

```
    public function getLoadedProductCollection()
    {
        return $this->_getProductCollection();
    }
```
The collection iteration inside the method in 2.4 causes the collection to load the db data so any after plugin or around plugin that calls the original method and tries to affect the collection does nothing to it in 2.4.

### Examples (*)

Let's consider this plugin.
```
public function afterGetLoadedProductCollection(
    \Magento\Catalog\Block\Product\ListProduct $subject, 
    \Magento\Eav\Model\Entity\Collection\AbstractCollection $collection
) {
    return $collection->addAttributeToFilter('some_attribute_code', 'some_value');
}
```
In 2.3 and before, this would cause the product list to filter only products that have the value `some_value` for the attribute with code `some_attribute_code` (beside the default filters).
In 2.4 this filter has no effect because the collection is already loaded when adding the filter.
### Proposed solution
Move the assign operation of category ids to products to the `_beforeToHtml` method in the same class after 
```
        if (!$collection->isLoaded()) {
            $collection->load();
        }
```
---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users with no workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @gabrieldagama 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-42162` were created

**Issue Available**: @gabrieldagama, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by gabrieldagama** [comment]
Hi @tzyganu. Thank you for your report. 
The issue has been fixed in magento/magento2#32390 by @mrtuvn in 2.4-develop branch
Related commit(s):
 - [c0d7665fe611981c0d651b0200bd577a2e333028](https://github.com/magento/magento2/commit/c0d7665fe611981c0d651b0200bd577a2e333028)

The fix will be available with the upcoming 2.4.3 release.

**Comment by mihaimatei** [comment]
> Hi @tzyganu. Thank you for your report. The issue has been fixed in #32390 by @mrtuvn in 2.4-develop branch Related commit(s):
> 
> * [c0d7665fe611981c0d651b0200bd577a2e333028](https://github.com/magento/magento2/commit/c0d7665fe611981c0d651b0200bd577a2e333028)
> 
> The fix will be available with the upcoming 2.4.3 release.

Fix was merged into 2.4-develop, but still not in any of the current 2.4.3 releases.

**Comment by hostep** [comment]
I can see that the code from https://github.com/magento/magento2/pull/32390 is included in Magento 2.4.4-beta4, so expect this to become available in 2.4.4 once it gets [in about a month](https://devdocs.magento.com/release/).

**Comment by Keivanph** [comment]
> So i was researching a problem with a template i'm using on one of my clients websites, and after some time i figured out it was this exact problem. The issue was that after upgrading to Magento 2.4.0 the pager seemed to do nothing anymore. It just did not work. It just loaded all products that are added to the category. You could see the pagination, you could click on the pages, but the products shown was always the same: All of them.
> 
> So i found out that this template has a affiliate marketing template in it, which loads the current loaded product collection via the function described. Once I removed that part, the pagination and everything started to function as normal again. Then i went on Google and started looking into this and found out this post on Git. Thanks a lot!
> 
> Question remains, is this patched in any recent version yet or is it still only available in the 2.4-develop branch? Is there a different way to patch this in? I know i can override core files but i don't really see that as a solution. Could bring up problems in the future.

Hi, out of curiosity what was that  "affiliate marketing template" ? I am facing very same problem with loading all of the products when ever I touch the collection in any level.
